libsodium-bindings/src/LibSodium/Bindings/KeyDerivation.hs

```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE CApiFFI #-}
+
+-- |
+-- Module: LibSodium.Bindings.KeyDerivation
+-- Description: Direct bindings to the key exchange functions implemented in Libsodium
+-- Copyright: (C) Hécate Moonlight 2022
+-- License: BSD-3-Clause
+-- Maintainer: The Haskell Cryptography Group
+-- Stability: Stable
+-- Portability: GHC only
+module LibSodium.Bindings.KeyDerivation
+  ( -- * Introduction
+    -- $introduction
+
+    -- ** Key Generation
+    cryptoKDFKeygen
+  , cryptoKDFDeriveFromKey
+
+    -- ** Constants
+  , cryptoKDFPrimitive
+  , cryptoKDFBytesMin
+  , cryptoKDFBytesMax
+  , cryptoKDFKeyBytes
+  , cryptoKDFContextBytes
+  )
+where
+
+import Foreign (Ptr, Word64, Word8)
+import Foreign.C (CChar, CInt (CInt), CSize (CSize), CUChar)
+
+foreign import capi "sodium.h crypto_kdf_keygen"
+  cryptoKDFKeygen
+    :: Ptr Word8
+    -- ^ Pointer that will hold the master key of length 'cryptoKDFKeyBytes'
+    -> IO ()
+
+foreign import capi "sodium.h crypto_kdf_derive_from_key"
+  cryptoKDFDeriveFromKey
+    :: Ptr CUChar
+    -- ^ Pointer that will hold the sub-key.
+    -> CSize
+    -- ^ Length of the sub-key.
+    -> Word64
+    -- ^ Identifier of the sub-key. Must not be reused for another sub-key.
+    -> Ptr CChar
+    -- ^ Pointer to the context, of size 'cryptoKDFContextBytes'.
+    -> Ptr CUChar
+    -- ^ Pointer to the master key, which will be of length 'cryptoKDFKeyBytes'.
+    -> IO CInt
+
+-- == Constants
+
+-- | The algorithm used by the key derivation algorithm, blake2b.
+--
+-- @since 0.0.1.0
+cryptoKDFPrimitive :: String
+cryptoKDFPrimitive = "blake2b"
```

Comment:
Is this something you're meant to pass into C? If so, `String` won't work, you need a C string.

Reply:
No, it's a fixed string on the C side but the C compiler is giving me shit about it.

```c

#define crypto_kdf_PRIMITIVE "blake2b"
SODIUM_EXPORT
const char *crypto_kdf_primitive(void)
            __attribute__ ((warn_unused_result));
```

I tried a lot of export types, and it was never quite right, so I gave up and just re-used the string.

Reply:
What exact error are you getting? A 'fixed string on the C side' is _nothing whatsoever_ like a Haskell `String`, so whatever the cause, your solution won't work.

Reply:
Does `Ptr CChar` not work?